### PR TITLE
[split/9+12] move udf, udaf, `Expr` to datafusion-expr module

### PR DESCRIPTION
**Pull request by Jimexist**
# Which issue does this PR close?

- follow up on #1784 
- part of #1750 

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion-expr/src/expr.rs [hunk 1]
```diff
@@ -0,0 +1,705 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (alamb) [inline datafusion-expr/src/expr.rs:1]: The big one!

File: datafusion/src/logical_plan/expr.rs [hunk 2]
```diff
@@ -21,379 +21,20 @@
 pub use super::Operator;
 use crate::error::{DataFusionError, Result};
 use crate::logical_plan::ExprSchemable;
-use crate::logical_plan::{window_frames, DFField, DFSchema};
-use crate::physical_plan::functions::Volatility;
-use crate::physical_plan::{aggregates, functions, udf::ScalarUDF, window_functions};
+use crate::logical_plan::{DFField, DFSchema};
```
Comment (alamb) [inline datafusion/src/logical_plan/expr.rs:24]: not much left in this module 👍

File: datafusion-expr/src/columnar_value.rs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+use arrow::array::ArrayRef;
+use arrow::array::NullArray;
+use arrow::datatypes::DataType;
+use arrow::record_batch::RecordBatch;
+use datafusion_common::ScalarValue;
+use std::sync::Arc;
+
+/// Represents the result from an expression
+#[derive(Clone)]
+pub enum ColumnarValue {
```
Comment (alamb) [inline datafusion-expr/src/columnar_value.rs:27]: I think `ColumnarValue` is really only used for expression evaluation and perhaps belongs in a physical crate (perhaps `physical_expr`). It is ok for now, but I don't think it belongs here long term

File: datafusion/src/physical_plan/aggregates.rs [hunk 4]
```diff
@@ -40,15 +40,6 @@ use expressions::{
 };
 use std::sync::Arc;
 
-/// the implementation of an aggregate function
```
Comment (alamb) [inline datafusion/src/physical_plan/aggregates.rs:43]: I like that you are slowly disentangling the physical from the logical plans

It seems like eventually we will have "physical plan" depending on Exprs and "logical plan" depending on exprs but not depending on each other 👍

**Review by alamb — APPROVED** [review]

File: datafusion/src/physical_plan/aggregates.rs (hunk 4 above)
Comment (Jimexist) [reply datafusion/src/physical_plan/aggregates.rs:43]: > I like that you are slowly disentangling the physical from the logical plans
> 
> It seems like eventually we will have "physical plan" depending on Exprs and "logical plan" depending on exprs but not depending on each other 👍

i had a look but i don't think it's happening anytime soon. now logical plan depends on table scan, which async scan to yield execution plan. it's hard to tear logican plan from execution plan given this chain.

**Pull request merged**